src/file_pipe_log/pipe_builder.rs

```diff
@@ -13,17 +14,74 @@ use crate::event_listener::EventListener;
 use crate::file_builder::FileBuilder;
 use crate::log_batch::LogItemBatch;
 use crate::pipe_log::{FileId, FileSeq, LogQueue};
+use crate::truncate::TruncateQueueParameter;
 use crate::Result;
 
 use super::format::FileNameExt;
 use super::log_file::{build_file_reader, LogFd};
 use super::pipe::{DualPipes, SinglePipe};
 use super::reader::LogItemBatchFileReader;
 
-pub trait ReplayMachine: Send + Default {
-    fn replay(&mut self, item_batch: LogItemBatch, file_id: FileId) -> Result<()>;
+pub trait ReplayMachine: Send + Default + Sync {
+    fn replay(&mut self, item_batch: LogItemBatch, file_id: FileId, path: &Path) -> Result<()>;
 
     fn merge(&mut self, rhs: Self, queue: LogQueue) -> Result<()>;
+
+    fn truncate<B: FileBuilder>(&mut self, _builder: Arc<B>) -> Result<()> {
+        Ok(())
+    }
+
+    fn init(&mut self, _params: Option<TruncateQueueParameter>) {}
+}
+
+pub trait ReplayMachineFactory: Sync {
+    type Machine: ReplayMachine;
+
+    fn new_machine(&self) -> Self::Machine;
+
+    fn parameters(&self) -> Option<TruncateQueueParameter>;
+
+    fn buildable(&self, _: LogQueue) -> bool {
+        true
+    }
+}
+
+#[derive(Clone)]
+pub struct ReplayMachineBuilder<M>
+where
+    M: ReplayMachine,
+{
+    pub(crate) truncate_params: Option<TruncateQueueParameter>,
+    _phantom: PhantomData<M>,
+}
+
+impl<M: ReplayMachine> Default for ReplayMachineBuilder<M> {
+    fn default() -> Self {
+        ReplayMachineBuilder {
+            truncate_params: None,
+            _phantom: Default::default(),
+        }
+    }
+}
```

Comment:
I don't think you understood https://github.com/tikv/raft-engine/pull/175#discussion_r772969268, only `TruncateMachine` has a special builder, other machines that can be constructed by `default` share one default builder.